File: Go-SDL2/shot.go
package main

import (
	"math"

	"github.com/veandco/go-sdl2/sdl"
)

type shot struct {
	_renderer    *sdl.Renderer
	_image       *sdl.Texture
	_rect        sdl.Rect
	_radius      float64
	_xPos        float64
	_yPos        float64
	_xVel        float64
	_yVel        float64
	_angle       float64
	_speed       float64
	_lifeTime    float64
	_lifeCounter float64
}

func newShot(renderer *sdl.Renderer, image *sdl.Texture, xPos float64, yPos float64, angle float64) (*shot, error) {
	s := &shot{}
	var err error

	s._renderer = renderer
	s._image = image
	s._angle = angle
	s._speed = 600
	s._xPos = xPos
	s._yPos = yPos
	s._lifeTime = 40
	s._lifeCounter = 0

	if _, _, s._rect.W, s._rect.H, err = s._image.Query(); err != nil {
		return s, err
	}

	s._radius = float64(s._rect.W / 2)

	radian := float64(s._angle) * math.Pi / 180
	s._xVel += math.Sin(radian) * s._speed
	s._yVel -= math.Cos(radian) * s._speed

	s._xPos -= float64(s._rect.W) / 2
	s._yPos -= float64(s._rect.H) / 2

	s._xPos += s._xVel * 0.05
	s._yPos += s._yVel * 0.05

	s._rect.X = int32(s._xPos + 0.5)
	s._rect.Y = int32(s._yPos + 0.5)

	return s, err
}

func (s *shot) left() int32 {
	return int32(s._xPos)
}

func (s *shot) right() int32 {
	return int32(s._xPos) + s._rect.W
}

func (s *shot) top() int32 {
	return int32(s._yPos)
}

func (s *shot) bottom() int32 {
	return int32(s._yPos) + s._rect.H
}

func (s *shot) center() (float64, float64) {
	x := s._xPos + float64(s._rect.W)/2
	y := s._yPos + float64(s._rect.H)/2
	return x, y
}

func (s *shot) radius() float64 {
	return s._radius
}

func (s *shot) angle() float64 {
	return s._angle
}

func (s *shot) alive() bool {
	return s._lifeCounter < s._lifeTime
}

func (s *shot) update(dt float64) {
	s._xPos += s._xVel * dt
	s._yPos += s._yVel * dt

	if s.right() < 0 {
		s._xPos = windowWidth
	} else if s.left() > windowWidth {
		s._xPos = -float64(s._rect.W)
	}

	if s.bottom() < 0 {
		s._yPos = windowHeight
	} else if s.top() > windowHeight {
		s._yPos = -float64(s._rect.H)
	}

	s._rect.X = int32(s._xPos + 0.5)
	s._rect.Y = int32(s._yPos + 0.5)

	s._lifeCounter += 60 * dt
}

func (s *shot) draw() error {
	err := s._renderer.CopyEx(s._image, nil, &s._rect, s._angle, nil, sdl.FLIP_NONE)

	return err
}

File: C-SDL2/shot.c
#include "shot.h"


bool shot_initialize(struct Shot **shot, SDL_Renderer *renderer, SDL_Texture *image) {
    *shot = calloc(1, sizeof(struct Shot));
    if (!*shot) {
        fprintf(stderr, "Error in calloc of new shot.\n");
        return true;
    }
    struct Shot *s = *shot;

    s->renderer = renderer;
    s->image = image;

    if (SDL_QueryTexture(s->image, NULL, NULL, &s->rect.w, &s->rect.h)) {
        fprintf(stderr, "Error while querying texture: %s\n", SDL_GetError());
        return true;
    }

    s->speed = 600;
    s->life_time = 40;
    s->shot_count = 0;
    s->alive = false;

    return false;
}

void shot_free(struct Shot *s) {
    if (s) {
        free(s);
    }
}

void shot_new(struct Shot *s, double x, double y, double angle, double radius) {
    s->angle = angle;
    double radian = s->angle * M_PI / 180;
    s->x_vel = sin(radian) * s->speed;
    s->y_vel = -cos(radian) * s->speed;
    s->x_pos = x - (s->rect.w / 2.0);
    s->y_pos = y - (s->rect.h / 2.0);
    s->x_pos += sin(radian) * radius;
    s->y_pos -= cos(radian) * radius;
    s->angle = angle;
    s->shot_count = 0;
    s->alive = true;
}

double shot_left(struct Shot *s) {
    return s->x_pos;
}

double shot_right(struct Shot *s) {
    return s->x_pos + s->rect.w;
}

double shot_top(struct Shot *s) {
    return s->y_pos;
}

double shot_bottom(struct Shot *s) {
    return s->y_pos + s->rect.h;
}

double shot_center_x(struct Shot *s) {
    return s->x_pos + s->rect.w / 2.0;
}

double shot_center_y(struct Shot *s) {
    return s->y_pos + s->rect.h / 2.0;
}

double shot_radius(struct Shot *s){
    return s->radius;
}

void shot_update(struct Shot *s, double dt) {
    if (s->alive) {
        s->shot_count += 60 * dt;
        if (s->shot_count > s->life_time) {
            s->alive = false;
            return;
        }

        s->x_pos += s->x_vel * dt;
        s->y_pos += s->y_vel * dt;

        if (s->x_pos + s->rect.w < 0) {
            s->x_pos = WINDOW_WIDTH;
        } else if (s->x_pos > WINDOW_WIDTH) {
            s->x_pos = -s->rect.w;
        }

        if (s->y_pos + s->rect.h < 0) {
            s->y_pos = WINDOW_HEIGHT;
        } else if (s->y_pos > WINDOW_HEIGHT) {
            s->y_pos = -s->rect.h;
        }

        s->rect.x = (int)(s->x_pos + 0.5);
        s->rect.y = (int)(s->y_pos + 0.5);
    }
}

bool shot_draw(struct Shot *s) {
    if (s->alive) {
        if ( SDL_RenderCopyEx(s->renderer, s->image, NULL, &s->rect, s->angle, NULL, SDL_FLIP_NONE)) {
            fprintf(stderr, "Error while rendering texture: %s\n", SDL_GetError());
            return true;
        }
    }
    return false;
}
